src/controllers/midi/midimessage.h

```diff
@@ -204,13 +210,35 @@ struct MidiInputMapping {
     }
 
     bool operator==(const MidiInputMapping& other) const {
-        return key == other.key && options == other.options &&
-                control == other.control && description == other.description;
+        return std::visit([&](auto first, auto second) {
+            using T = std::decay_t<decltype(first)>;
+            using U = std::decay_t<decltype(second)>;
+
+            if constexpr (!std::is_same_v<T, U>) {
+                return false;
+            } else if constexpr (std::is_same_v<T, ConfigKey>) {
+                return key == other.key &&
+                        options == other.options &&
+                        std::get<ConfigKey>(control) == std::get<ConfigKey>(other.control) &&
+                        description == other.description;
+            } else if constexpr (std::is_same_v<T, std::shared_ptr<QJSValue>>) {
+                const std::shared_ptr<QJSValue> otherControl =
+                        std::get<std::shared_ptr<QJSValue>>(other.control);
+                const std::shared_ptr<QJSValue> thisControl =
+                        std::get<std::shared_ptr<QJSValue>>(control);
+                return key == other.key && options == other.options &&
+                        otherControl->strictlyEquals(*thisControl) &&
+                        description == other.description;
+            } else
+                static_assert(always_false_v<T>, "non-exhaustive visitor");
+        },
+                control,
+                other.control);
     }
 
     MidiKey key;
     MidiOptions options;
-    ConfigKey control;
+    std::variant<ConfigKey, std::shared_ptr<QJSValue>> control;
```

Comment:
> We may consider a new name, "destination" or such...
> 
> Is it correct, that it can now carry a
> 
>     * ConfigKey form a Mixxx control object
> 
>     * Group + Name of a script binding comming from XML
> 
>     * The new function pointer.
> 
> 


Yes, though group+name of script binding was previously also stored in the same configkey. It's a bad design decision made way earlier. Refactoring this now is out-of-scope IMO.

> I think we should also get rid of this on the fly wrapping that is done for every incoming midi message:

yes, I agree. this PR would make that possible now. Though it can also be done later.

Reply:
Before, It was a ConfigKey called `control`. Now this `control` is wrapped in a std::variant witch really requires a new name to emphasize that and not create confusion when reading it elsewhere. 

I am fine with going in small steps here. Than please add a TODO comment for adding a third type for more type safety in std::variant  and another TODO for getting rid of the on the fly wrapping. I am afraid otherwise this discussion will got into oblivion.